### PR TITLE
[NFC] Add a more powerful API for collecting heap types

### PR DESCRIPTION
**Pull request by tlively**
Many passes need to know both the set of all used types and also the
sets of private or public types. Previously there was no API to get both
at once, so getting both required two API calls that internally
collected all the types twice.

Furthermore, there are many reasons to collect heap types, and they have
different requirements about precisely which types need to be collected.
For example, in some edge cases the IR can reference heap types that do
not need to be emitted into a binary; passes that replace all types
would need to collect these types, but the binary writer would not. The
existing APIs for collecting types did not distinguish between these use
cases, so the code conservatively collected extra types that were not
always needed.

Refactor the type collecting code to expose a new API that takes a
description of which types need to be collected and returns the
appropriate types, their use counts, and optionally whether they are
each public or private.

Keep this change non-functional by commenting on places where the code
could be cleaned up or improved rather than actually making the changes.
Follow-up PRs will implement the improvements, which will necessarily
come with test changes.

### CODE REVIEW

File: src/ir/module-utils.h [hunk 1]
```diff
@@ -442,6 +443,22 @@ template<typename T> struct CallGraphPropertyAnalysis {
   }
 };
 
+enum class TypeInclusion { AllTypes, UsedIRTypes, BinaryTypes };
+
+enum class ClassifyAction { NoClassify, Classify };
+
+enum class TypeClassification { Unknown, Public, Private };
```
Comment (kripken) [inline src/ir/module-utils.h:450]: Please add comments for these enum values (e.g. UsedIRTypes and Classify are not trivial).

File: src/ir/module-utils.cpp [hunk 2]
```diff
@@ -339,142 +339,143 @@ struct Counts {
     assert(sig.params.size() == 0);
     if (sig.results.isTuple()) {
       // We have to use a function type.
-      controlFlowSignatures[sig]++;
+      ++controlFlowSignatures[sig];
     } else if (sig.results != Type::none) {
       // The result type can be emitted directly instead of using a function
       // type.
-      note(sig.results[0]);
+      note(sig.results);
```
Comment (kripken) [inline src/ir/module-utils.cpp:346]: Is this a bugfix?

File: src/ir/module-utils.cpp [hunk 3]
```diff
@@ -537,44 +543,53 @@ InsertOrderedMap<HeapType, size_t> getHeapTypeCounts(Module& wasm,
...
     }
     // All the rec group members are public as well.
+    bool inserted = false;
     for (auto member : type.getRecGroup()) {
-      if (!publicTypes.insert(member)) {
-        // We've already inserted this rec group.
-        break;
+      if (auto it = types.find(member); it != types.end()) {
+        if (it->second.classification == TypeClassification::Public) {
+          // We've already inserted this rec group.
```
Comment (kripken) [inline src/ir/module-utils.cpp:584]: I find these two lines hard to follow. Why does the classification being public show that we've already inserted the rec group? Perhaps the comment should be more like "we've already inserted this rec group, and marked it as public"?

File: src/ir/module-utils.cpp (hunk 2 above)
Comment (tlively) [reply src/ir/module-utils.cpp:346]: No, this is NFC. We allow indexing into all types for uniformity, but the only valid index for a non-tuple is 0 and the type at index 0 is the original type itself. The indexing was useless extra work here.

Comment (kripken) [reply src/ir/module-utils.cpp:346]: Sorry, I don't follow. `sig.results` might be a tuple if the signature is a multivalue return, like `(i32 f64)`. Before we just read `i32` in that case?

Comment (tlively) [reply src/ir/module-utils.cpp:346]: We handle the tuple case separately above. At this point the type is known to not be a tuple.

Comment (kripken) [reply src/ir/module-utils.cpp:346]: Ah, thanks, that's what I was missing.

**Comment by tlively** [comment]
Comments added and improved!

File: src/ir/module-utils.h [hunk 4]
```diff
@@ -442,6 +443,35 @@ template<typename T> struct CallGraphPropertyAnalysis {
   }
 };
 
+// Which types to collect.
+//
+//   AllTypes - Any type anywhere reachable from anything
```
Comment (kripken) [inline src/ir/module-utils.h:448]:
```suggestion
//   AllTypes - Any type anywhere reachable from anything.
```

File: src/ir/module-utils.h [hunk 5]
```diff
@@ -442,6 +443,35 @@ template<typename T> struct CallGraphPropertyAnalysis {
...
+//   UsedIRTypes - Same as AllTypes, but excludes types reachable only because
+//   they are in a rec group with some other used type and types that are only
+//   used from other unreachable types.
+//
+//   BinaryTypes - Only types that need to appear in the module's type section.
+//
+enum class TypeInclusion { AllTypes, UsedIRTypes, BinaryTypes };
+
+// Whether to classify collected types as public and private.
+enum class ClassifyAction { NoClassify, Classify };
```
Comment (kripken) [inline src/ir/module-utils.h:459]: Classify is fairly generic. Maybe the public/private property could be "publicity" or "exposure"? (or the C++ term would be "access" I guess). Then this could be `PublicityHandling`.

File: src/ir/module-utils.h [hunk 6]
```diff
@@ -442,6 +443,35 @@ template<typename T> struct CallGraphPropertyAnalysis {
...
+
+// Whether a type is public or private. If no classification occurs, the type
+// will be Unknown instead.
+enum class TypeClassification { Unknown, Public, Private };
```
Comment (kripken) [inline src/ir/module-utils.h:463]: If the above comment makes sense to you, then `TypeClassification` could simply be `Publicity`.

Comment (tlively) [reply src/ir/module-utils.h:463]: I will use `Visibility`.

**Review by kripken — APPROVED** [review]
